### PR TITLE
[ci-visibility] Test suite visibility: support `jest`

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
Add test suite level visibility to `jest`.

### Motivation
Increase visibility for customers.

### Plugin Checklist
- [x] Unit tests.

### CODE REVIEW

File: packages/datadog-plugin-jest/src/index.js [hunk 1]
```diff
@@ -58,6 +64,89 @@ class JestPlugin extends Plugin {
     this.testEnvironmentMetadata = getTestEnvironmentMetadata('jest', this.config)
     this.codeOwnersEntries = getCodeOwnersFileEntries()
 
+    this.addSub('ci:jest:session:start', (command) => {
+      if (!this.config.isAgentlessEnabled) {
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:68]: Why is this needed for the suite spans but not the test spans?

File: packages/datadog-plugin-jest/src/index.js [hunk 2]
```diff
@@ -58,6 +64,89 @@ class JestPlugin extends Plugin {
...
+        return
+      }
+      const store = storage.getStore()
+      const childOf = getTestParentSpan(this.tracer)
+      const testSessionSpanMetadata = getTestSessionCommonTags(command, this.tracer._version)
+
+      this.command = command
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:75]: This should be stored in some sort of the context instead of on the plugin instance.

File: packages/datadog-plugin-jest/test/circus.spec.js [hunk 3]
```diff
@@ -312,6 +279,98 @@ describe('Plugin', function () {
           )
         })
       }
+
+      describe('agentless', () => {
```
Comment (rochdev) [inline packages/datadog-plugin-jest/test/circus.spec.js:283]: What is the distinction between agentless and not agentless? I thought the feature needed to always be agentless in the first place?

File: packages/datadog-instrumentations/src/jest.js [hunk 4]
```diff
@@ -72,6 +82,14 @@ function getWrappedEnvironment (BaseEnvironment) {
       this.testSuite = getTestSuitePath(context.testPath, rootDir)
       this.nameToParams = {}
       this.global._ddtrace = global._ddtrace
+
+      if (config.projectConfig && config.projectConfig.testEnvironmentOptions) { // newer versions
+        this._ddTestSessionId = config.projectConfig.testEnvironmentOptions._ddTestSessionId
```
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-instrumentations/src/jest.js:87]: Also, why is this passed to the instrumentation in the first place? It should have no knowledge of tracing.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/jest.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/jest.js:87]: this config is the only info that is passed to the worker thread that is spawned to run the tests

File: packages/datadog-plugin-jest/src/index.js (hunk 1 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:68]: session and suite events are only processed by the agentless intake. Tests are processed by both agent and agentless intake

File: packages/datadog-plugin-jest/test/circus.spec.js (hunk 3 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/test/circus.spec.js:283]: this is just for convenience: in this describe we have tests that require agentless mode. This is so that we can easily identify them and modify the inputs to `agent.load`

File: packages/datadog-instrumentations/src/jest.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-instrumentations/src/jest.js:87]: Could this be done without passing the data to the instrumentation? The instrumentation should never know about tracing, otherwise we can never upstream the changes, and it's also bad practice generally speaking for the publisher to have to know about its subscriber.

[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/jest.js:87]: can you think of anything? We might be able to pass some environment variables to the worker threads that change the way dd-trace is initialised in them? I'm not sure if that's even possible and it sounds risky, as we may inadvertently affect processes we don't want to affect

[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-instrumentations/src/jest.js:87]: How does Jest connect them in the first place?

[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/jest.js:87]: the main processes starts a bunch of children processes and tells them which test files they need to run

[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-instrumentations/src/jest.js:87]: What I mean is how does the communication happen between the main process and the workers? How does the main process identifies what is done by which worker for example?

[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/jest.js:87]: they have cross process communication. The child reports its results to the parent: https://github.com/facebook/jest/blob/4def94b073cad300e99de378ba900e6ba9b7032f/packages/jest-worker/src/workers/processChild.ts#L71-L77

[the remark this answers is not in the document]
Comment (rochdev) [reply packages/datadog-instrumentations/src/jest.js:87]: Any way that we could link things from the information available there? If it's really not possible at all, any way to share something between the 2 processes that would be more realistic to add to Jest in the future? So basically something not related to Datadog.

File: packages/datadog-plugin-jest/src/index.js [hunk 5]
```diff
@@ -58,6 +64,88 @@ class JestPlugin extends Plugin {
...
+      if (!this.config.isAgentlessEnabled) {
+        return
+      }
+
+      const { _ddTestSessionId: testSessionId, _ddTestCommand: testCommand } = testEnvironmentOptions
+
+      const store = storage.getStore()
+
+      const testSessionSpanContext = this.tracer.extract('text_map', {
+        'x-datadog-trace-id': id(testSessionId).toString(10),
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:120]: Why not store the ID in the same format it will be used here instead of converting from hex to int?

File: packages/datadog-plugin-jest/src/index.js [hunk 6]
```diff
@@ -58,6 +64,88 @@ class JestPlugin extends Plugin {
...
+        'x-datadog-span-id': id(testSessionId).toString(10),
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:121]: What is `x-datadog-span-id`?

File: packages/datadog-plugin-jest/src/index.js [hunk 7]
```diff
@@ -58,6 +64,88 @@ class JestPlugin extends Plugin {
...
+    // Test suites can be run in a different process from jest's main one.
+    // This subscriber changes the configuration objects from jest to inject the trace id
+    // of the test session to the processes that run the test suites.
+    this.addSub('ci:jest:session:configuration', configs => {
+      if (!this.config.isAgentlessEnabled) {
+        return
+      }
+      const testSessionSpan = storage.getStore().span
+      configs.forEach(config => {
+        config._ddTestSessionId = testSessionSpan.context()._traceId.toString(16)
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:105]: Could these be stored outside of a variable from Jest like in a WeakMap?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-plugin-jest/src/index.js (hunk 6 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:121]: yeah this slipped. Removing it

File: packages/datadog-plugin-jest/src/index.js (hunk 5 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:120]: yeah sure, good catch

Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js]: done in https://github.com/DataDog/dd-trace-js/pull/2122/commits/24a2f1bcce959a6158ae71c9a49475c80b55a147

File: packages/datadog-plugin-jest/src/index.js (hunk 6 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js]: done in https://github.com/DataDog/dd-trace-js/pull/2122/commits/24a2f1bcce959a6158ae71c9a49475c80b55a147

File: packages/datadog-plugin-jest/src/index.js (hunk 7 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js]: sadly no. This is how we inject this id to the worker processes that run the tests. Storing in a weak map here would do nothing, because it would be empty in the worker process

**Review by rochdev — APPROVED** [review]
